app/frontend/src/components/ImageInput/ImageInput.tsx

```diff
@@ -49,23 +63,31 @@ const useStyles = makeStyles((theme) => ({
   },
 }));
 
-interface AvatarInputProps {
+type ImageInputProps = {
   className?: string;
   control: Control;
   id: string;
   initialPreviewSrc?: string;
   name: string;
-  userName?: string;
-}
+};
 
-export function AvatarInput({
-  className,
-  control,
-  id,
-  initialPreviewSrc,
-  name,
-  userName,
-}: AvatarInputProps) {
+type AvatarInputProps = {
+  type: "avatar";
+  userName: string;
+};
+
+type SquareImgInputProps = {
+  type: "rect";
+  alt: string;
+  grow?: boolean;
+  height?: number;
+  width?: number;
+};
+
+export function ImageInput(
+  props: (AvatarInputProps | SquareImgInputProps) & ImageInputProps
```

Comment:
It would simplify things a bit here if we have all of these as `interfaces`, then have `AvatarInputProps` and `SquareImgInputProps` extend from `ImageInputProps`. That way, in the `ImageInput` props type, you don't need the `& ImageInputProps` bit, which I think makes it a little bit more intuitive to read

Reply:
I did it that way originally, but I couldn't get it to infer the type once I branched on `type`

Reply:
Still works fine for me according to VS Code, considering the `type` field you got there is exactly what TS used to discriminate the type.

type before the if ternary branch:
<img width="587" alt="Screenshot 2021-04-21 at 01 56 35" src="https://user-images.githubusercontent.com/13669362/115481788-450f3600-a245-11eb-9d07-13b1880902b3.png">

`SquareImgInputProps` correctly narrowed in the else ternary branch:
<img width="623" alt="Screenshot 2021-04-21 at 01 56 47" src="https://user-images.githubusercontent.com/13669362/115481794-46d8f980-a245-11eb-9b6d-a6e136f4d705.png">